Overview.bs

```diff
@@ -623,13 +626,479 @@ The algorithm:
+  <tr>
+    <td>uint16</td>
+    <td><dfn for="Format 2 Patch Map">uriTemplateLength</dfn></td>
+    <td>
+      Length of the uriTemplate string.
+    </td>
+  </tr>
+  <tr>
+    <td>uint8</td>
+    <td><dfn for="Format 2 Patch Map">uriTemplate</dfn>[uriTemplateLength]</td>
+    <td>
+      A [[!UTF-8]] encoded string. Contains a [[#uri-templates]] which is used to produce URIs associated with each entry.
+    </td>
+  </tr>
+</table>
+
+<dfn>Mapping Entries</dfn> encoding:
+
+<table>
+  <tr>
+    <th>Type</th><th>Name</th><th>Description</th>
+  </tr>
+  <tr>
+    <td>[=Mapping Entry=]</td>
+    <td><dfn for="Mapping Entries">entries</dfn>&lsqb;[=Format 2 Patch Map/entryCount=]&rsqb;</td>
+    <td>List of entries. Each entry has a variable length, which is determined following [$Interpret Format 2 Patch Map Entry$].</td>
+  </tr>
+</table>
+
+<dfn>Mapping Entry</dfn> encoding:
+
+<table>
+  <tr>
+    <th>Type</th><th>Name</th><th>Description</th>
+  </tr>
+  <tr>
+    <td>uint8</td>
+    <td><dfn for="Mapping Entry">format</dfn></td>
+    <td>
+      A bit field. Bit 0 (least significant bit) through bit 5 indicate the presence of optional fields. If bit 6 is set this entry is
+      ignored. Bit 7 is reserved for future use and set to 0.
+    </td>
+  </tr>
+
+  <tr>
+    <td>uint8</td>
+    <td><dfn for="Mapping Entry">featureCount</dfn></td>
+    <td>
+      Number of feature tags in the featureTags list. Only present if [=Mapping Entry/format=] bit 0 is set.
+    </td>
+  </tr>
+  <tr>
+    <td>Tag</td>
+    <td><dfn for="Mapping Entry">featureTags</dfn>[featureCount]</td>
+    <td>
+      List of [[open-type/featuretags|feature tags]] in the entry's [=font subset definition=]. Only present if
+      [=Mapping Entry/format=] bit 0 is set.
+    </td>
+  </tr>
+
+  <tr>
+    <td>uint8</td>
+    <td><dfn for="Mapping Entry">designSpaceCount</dfn></td>
+    <td>
+      Number of elements in the design space list. Only present if [=Mapping Entry/format=] bit 0 is set.
+    </td>
+  </tr>
+  <tr>
+    <td>[=Design Space Segment=]</td>
+    <td><dfn for="Mapping Entry">designSpaceSegments</dfn>[designSpaceCount]</td>
+    <td>
+      List of design space segments in the entry's [=font subset definition=]. Only present if [=Mapping Entry/format=] bit 0 is set.
+    </td>
+  </tr>
+
+  <tr>
+    <td>uint8</td>
+    <td><dfn for="Mapping Entry">copyCount</dfn></td>
+    <td>
+      Number of entries in the copyIndices list. Only present if [=Mapping Entry/format=] bit 1 is set.
+    </td>
+  </tr>
+  <tr>
+    <td>uint16</td>
+    <td><dfn for="Mapping Entry">copyIndices</dfn>[copyCount]</td>
+    <td>
+      List of indices from the [=Mapping Entries/entries=] array whose [=font subset definition=] should be copied into this entry. All
+      values must be less than the index of this [=Mapping Entry=] in [=Mapping Entries/entries=]. Only present if
+      [=Mapping Entry/format=] bit 1 is set.
+    </td>
+  </tr>
+
+  <tr>
+    <td>int16</td>
```

Comment:
So is the idea here is that you can have an effective entry id greater than 2^16? Will that work with every needed combination? Do we need to push the compression that far? Or is it that you want sub-sequences of the records to stay consistent if there are earlier insertions? Given that the absolute entry IDs seem to have real meaning I'm not sure that would make sense.

Also, this description may not be entirely wrong given the vagueness of "produce" but it is misleading, given that the id will be [previous entry id] + [entryIdDelta] + 1. 

(Overall I'm asking why you can't just specify the entry ID with this field, making it be one plus the previous ID if the field is missing.)

Reply:
Two reasons I see the delta being useful:
1. in the shared brotli case applying a patch will cause all of the remaining entries to get new id numbers further down in the space, so the delta can be used to jump the first entry id ahead and the rest will increment by one as normal.
2. We have the ability to have mapping entries that are copied into others but set the ignore bit so they themselves aren't used. Deltas could be used to keep the non-ignored entries having sequential ids. Though thinking about this, maybe I should specify that ignored entries don't have entry ids and thus don't cause the normal +1 increment.

I don't see a need to have >2^16 id's, but there's also no reason they wouldn't work as things currently are. Though we probably should set an upper bound somewhere (like how I disallow negative values).

Reply:
I'll clarify the description to include the "+1". That's further clarified later on as well in the interpreting section, but might as well keep this consistent.

Reply:
I think omitting the IDs for ignored entries makes sense.

> in the shared brotli case applying a patch will cause all of the remaining entries to get new id numbers further down in the space, so the delta can be used to jump the first entry id ahead and the rest will increment by one as normal.

So the idea is that given that the segmentation is fixed across the whole graph, portions of the map for a given part of the segmentation will tend to repeat at different nodes on the graph, and the delta helps with that? I can see how that will work in practice with the implicit + 1 when the field is missing, I'm more fuzzy on it working when you need to specify the field in the first place. I guess the idea is that what a patch does is remove an entry, and the delta at that point will need to change but the other deltas already established from earlier patches won't need to?

Reply:
On second thought we'll want to keep assigning IDs to ignored. When applying glyph keyed patches we want there to be a mechanism where the client can remove entries without modifying the length of the IFT/IFTX tables. The ignored bit allows us to do that in format 2, but only if we keep assigning ignored entries IDs (so ids for following entries aren't changed).

To give a concrete example if we start with a subset which has patches 1...n. After applying one of those patches all patches 1...n no longer can be applied since the base has changed, and thus mapping will be updated to instead link to a new set of patches with ids m+1 ... m + n - 1 (where m depends on the patch applied). The key side (subset def) of the mapping entries remains the same, but all of the ID's need to be shifted. The first entry get's a delta of +m, the applied entry is removed, and all remaining entries are left untouched and use the implicit +1.

Also depending on how an encoder is implemented it may choose for some reason not to use completely sequential id's. Having the deltas allows for arbitrary ID's sequences to be assigned to each entry so the encoder isn't restricted in choosing ids.

Reply:
> (Overall I'm asking why you can't just specify the entry ID with this field, making it be one plus the previous ID if the field is missing.)

On re-reading this I think I missed this part. Setting the id value directly should also work fine for the ways I typically see this being used. For most cases I believe the encoded size would be the same using deltas or absolute values.

I still have a slight preference for using deltas as it does work nicely in a case where you have a non-sequential list ID's that you want to shift uniformly with a patch application (the current encoder prototype doesn't make use of this, but maybe others might?).

I've also been thinking we probably want to increase the size of this field to at least 24 bits (whether it's a delta or absolute value) and the entry count to 32 bits to match format 1.

Reply:
I think that question from the first response is resolved for me. As I understand it the idea of using the delta instead of the value is that subsequent patches to earlier portions of the range will need to change the delta in fewer cases.

Reply:
Going to merge this as is for now to unblock some follow on changes, but I'll follow up with another PR to adjust the field sizes, and if needed we can continue the discussion on absolute vs delta ids there.